### ISSUE TITLE
Tracking issue for end-to-end wasi-threads support

### ISSUE THREAD

**Issue by abrown** [WebAssembly/wasi-threads#10]
Several of us (@loganek, @sunfishcode, @sunfishcode, @haraldh) have been working towards implementing all of the pieces to demonstrate an end-to-end `wasi-threads` example. The current direction is to implement this in Wasmtime, though @loganek has also opened a PR to do so in WAMR ([#1638](https://github.com/bytecodealliance/wasm-micro-runtime/pull/1638)). To that end, this issue is meant to track all of the various parts needed not only to show a proof-of-concept, but to upstream enough code for (fearless) users to try out this new functionality. I do not expect this to be a comprehensive plan, but only to implement what is necessary for the "stage 1" functionality described by @alexcrichton [here](https://github.com/bytecodealliance/wasmtime/pull/4949#issuecomment-1282975456) (i.e., no component model integration).

We can split this into areas and I've made an effort to try to order the tasks within these.

### Specification

-   [ ] spec: consider adding `wasi_thread_exit` for early return from a thread ([#7](https://github.com/WebAssembly/wasi-threads/issues/7)); will require some implementation in `wasi-libc` and `wasmtime`
-   [x] spec: fix CI issues ([#26](https://github.com/WebAssembly/wasi-threads/pull/26))

### Toolchains

-   [x] toolchain: allow importing and exporting the same memory in `wasm-ld`; done in ~~[D135898](https://reviews.llvm.org/D135898)~~
-   [x] toolchain: allow compiling programs with `wasm32-wasi-threads` (~~[#326](https://github.com/WebAssembly/wasi-libc/issues/326)~~); in progress at ~~[#331](https://github.com/WebAssembly/wasi-libc/pull/331)~~ and [#274](https://github.com/WebAssembly/wasi-sdk/pull/274)
-   [ ] toolchain: teach the toolchains to `import` shared memories by default; it seems most natural to have both the parent and child threads import a memory versus exporting it [^create-shared]; under discussion at [#502](https://github.com/WebAssembly/WASI/issues/502)

[^create-shared]: this also likely involves figuring out a better place to create the shared memory initially in Wasmtime (instead of [here](https://github.com/bytecodealliance/wasmtime/compare/main...haraldh:wasi-threads-working#diff-d213ee8ac09e5c4d18a2ac5faddc59fe1a5916f583693fbc45464e9b23dbd059R126))

### Libraries

-   [x] wasi-libc: only run C constructors once ([#339](https://github.com/WebAssembly/wasi-libc/pull/339)); ready to merge
-   [x] wasi-libc: initial thread-local values must be `memcpy`-ed to the TLS area; this seems to be handled by some combination of [#342](https://github.com/WebAssembly/wasi-libc/pull/342) and ~~[#343](https://github.com/WebAssembly/wasi-libc/pull/343)~~
-   [x] wasi-libc: resolve spinlock questions to merge pthread spinlock support ([#324](https://github.com/WebAssembly/wasi-libc/pull/324))
-   [ ] wasi-libc: create a way to run [`libc-test` tests](http://nsz.repo.hu/git/?p=libc-test;a=blob;f=src/functional/pthread_mutex.c;h=1c048aefeec6b9bde7f2f15b3b49e5e304dfad6a;hb=HEAD); in progress at [#369](https://github.com/WebAssembly/wasi-libc/pull/369)
-   [ ] wasi-libc: audit for any missing pthreads APIs &mdash; the fundamentals are there but we need to determine what functions remain to be enabled

### Engines

-   [x] wasmtime: implement `wait` and `notify`; on Linux, this could use `futex` as done [here](https://github.com/bytecodealliance/wasmtime/compare/main...haraldh:wasi-threads-working#diff-aeecd191ed2832ecc670202be8de1d73ff30c8476f710bc5aa6d759bd5d925b9R434) by @haraldh but we need a solution for all OSes (also, is the [32-bit limitation](https://man7.org/linux/man-pages/man2/futex.2.html) a problem?)
-   [x] wasmtime: avoid mutation of buffers/strings passed in to host functions by Wiggle; [#5054](https://github.com/bytecodealliance/wasmtime/pull/5054), [#5225](https://github.com/bytecodealliance/wasmtime/pull/5225), [#5229](https://github.com/bytecodealliance/wasmtime/pull/5229) [#5264](https://github.com/bytecodealliance/wasmtime/pull/5264), [#5268](https://github.com/bytecodealliance/wasmtime/pull/5268)
-   [x] wasmtime: fix `validate_atomic_addr`; ~~[#5063](https://github.com/bytecodealliance/wasmtime/pull/5063)~~ ~~[#5239](https://github.com/bytecodealliance/wasmtime/pull/5239)~~
-   [x] wasmtime: https://github.com/bytecodealliance/wasmtime/issues/5235
-   [x] wasmtime: implement `wasi-threads`; in progress at [#5484](https://github.com/bytecodealliance/wasmtime/pull/5484)

I'm completely open adding/removing/editing the items above as well as moving this issue somewhere else but I felt it would be helpful to keep track of the state of things.

**Comment by haraldh** [comment]
> wasi-libc: initial thread-local values must be `memcpy`-ed to the TLS area; this seems to be handled by some combination of [#342](https://github.com/WebAssembly/wasi-libc/pull/342) and [#343](https://github.com/WebAssembly/wasi-libc/pull/343) (need to resolve the overlap)

This is done in #342 and #343 with `__wasm_init_tls(new_tls_mem)` where `__wasm_init_tls` is linker generated.

**Comment by haraldh** [comment]
> * wasmtime: implement `wait` and `notify`; on Linux, this could use `futex` as done [here](https://github.com/bytecodealliance/wasmtime/compare/main...haraldh:wasi-threads-working#diff-aeecd191ed2832ecc670202be8de1d73ff30c8476f710bc5aa6d759bd5d925b9R434) by @haraldh but we need a solution for all OSes (also, is the [32-bit limitation](https://man7.org/linux/man-pages/man2/futex.2.html) a problem?)

So, there is the [atomic-wait](https://crates.io/crates/atomic-wait) crate written by Mara Bos. Here is what she wrote about the other OSes:

> 
> Cross platform atomic wait and wake (aka futex) functionality.
> 
> This crate only supports functionality that's available on all of Linux, Windows, and macOS. That is:
> 
> * Only AtomicU32 is supported. (Linux currently only supports 32-bit futexes.)
> * Only the "wait", "wake one", and "wake all" operations are supported. (Linux supports more operations, but Windows and macOS don't.)
> * No timeouts. (macOS doesn't have a stable/public API for timeouts.)
> * The wake operations don't return the number of threads woken up. (Only Linux supports this.)
> 

So I don't know what OS the person used, whoever specified `memory.atomic.notify`, `memory.atomic.wait32` and `memory.atomic.wait64`.

**Comment by haraldh** [comment]
`wasi_thread_kill` would also be a good extension

**Comment by sbc100** [comment]
> `wasi_thread_kill` would also be a good extension

Do you mean `pthread_exit`, or something to implement `pthread_cancel` or something to implement pthread_kill (which is oddly names since it delivers signals and doesn't control the lifetime of a thread).

**Comment by haraldh** [comment]
> > `wasi_thread_kill` would also be a good extension
> 
> Do you mean `pthread_exit`, or something to implement `pthread_cancel` or something to implement pthread_kill (which is oddly names since it delivers signals and doesn't control the lifetime of a thread).

Yeah, something to implement `pthread_cancel` I had in mind.

**Comment by sbc100** [comment]
So there are two types of pthread cancellation: DEFERRED and ASYNCHRONOUS : https://man7.org/linux/man-pages/man3/pthread_setcanceltype.3.html.

IIUC deferred cancellation is much simpler and doesn't need any OS support, its all implementable in user space.  Async cancellation is much harder and normal relies on some kind of async signal primitive under the hood.  I have previously proposed that we simply ignore asynchronous cancellation since async signal very different to anything that exists in wasi/wasm today.  I would also argue that very few real world program depend on it, but my information here is obviously not complete.

**Comment by abrown** [comment]
> wasmtime: implement wait and notify; on Linux, this could use futex as done [here](https://github.com/bytecodealliance/wasmtime/compare/main...haraldh:wasi-threads-working#diff-aeecd191ed2832ecc670202be8de1d73ff30c8476f710bc5aa6d759bd5d925b9R434) by @haraldh but we need a solution for all OSes (also, is the [32-bit limitation](https://man7.org/linux/man-pages/man2/futex.2.html) a problem?)

@haraldh, it looks like V8 bottoms out in their own `FutexEmulation` [implementation](https://github.com/v8/v8/blob/c80394dd3c9e651574359afc2db61293aa1ad2a5/src/execution/futex-emulation.h#L20-L26), which is used for all platforms, even Linux.

**Comment by abrown** [comment]
I've created https://github.com/bytecodealliance/wasmtime/pull/5274 to make a way to measure how much (or little) locking will affect WASI performance.

**Comment by turbolent** [comment]
Once spec or some form of docs/description, and some example/test binaries are available, I'd love to implement support in https://github.com/turbolent/w2c2.

**Comment by yamt** [comment]
in this weekend, i implemented wasi-threads for [toywasm](https://github.com/yamt/toywasm).
while it still has rough edges, it can run toywasm itself built for wasi, which uses pthread with the latest wasi-sdk.

**Comment by turbolent** [comment]
@loganek Thank you, I'll give that a try 👍 
@yamt Congrats, great work!

**Comment by loganek** [comment]
FYI WAMR implementation is being tracked here: https://github.com/bytecodealliance/wasm-micro-runtime/issues/1790

**Comment by abrown** [comment]
For those interested, I updated the description for this issue with the state of the latest PRs. With a highly-customized environment (e.g., https://github.com/bytecodealliance/wasmtime/pull/5484), I have a benchmark that runs parallel compression using wasi-threads in Wasmtime. I think @yamt has also shown how to run ffmpeg in parallel somewhere (with Wasmtime or his toy engine, I'm not sure).

In my view, the current state of this issue is that the main check boxes above are all covered by applicable PRs. We simply need to clean up and merge those PRs for wasi-threads to be generally available for testing. It might take a while for things to start showing up in various releases and I would expect some bugs to be found in various projects, but my impression is that this effort is almost complete.

**Comment by loganek** [comment]
Should we target including the API in preview2? I don't mean to rush, but feels like we already are confident about the API and most of the important ambiguities have been clarified.

**Comment by abrown** [comment]
@loganek, that's a good idea. @sunfishcode, any thoughts on that?

**Comment by sunfishcode** [comment]
Because of wasi-thread's decision to go forward with instance-per-thread, I myself don't know how this could be be possible within any predictable timeframe.

**Comment by abrown** [comment]
I am going to close this issue since the main idea of it &mdash; tracking the implementation of wasi-threads end-to-end &mdash; is now upstreamed in both Wasmtime and WAMR and all of the toolchain work to get here is contained in the [`wasi-sdk-20+threads`](https://github.com/WebAssembly/wasi-sdk/releases/tag/wasi-sdk-20%2Bthreads) pre-release of wasi-sdk. As you may notice, not all of the checkboxes in the original issue are checked but this is fine: over the last few months things have progressed in ways that make some of the items unnecessary and the necessary ones to continue on as separate issues or PRs. I tried to describe all of the work done here (by me and others!) in this blog post: https://bytecodealliance.org/articles/wasi-threads. Thanks to everyone who contributed!
